server-phoenix/lib/helios_web/plugs/user_authentication.ex

```diff
@@ -0,0 +1,80 @@
+defmodule HeliosWeb.Plugs.UserAuthentication do
+  import Plug.Conn
+  alias Helios.User
+  alias Helios.Repo
+
+  def init(default), do: default
+
+  def call(conn, _default) do
+    # To verify that data was signed by a particular party, you must have the following information:
+    #
+    #    The public key of the party that signed the data.
+    user_public_key = get_user_name(conn) |> get_public_key(conn)
+
+    #
+    #    The digital signature.
+    [encoded_signature] = get_signature(conn)
+
+    #
+    #    The data that was signed.
+    signed_data = conn.assigns.raw_body
+
+    #
+    #    Now, authenticate the signature:
+    case Base.decode64(encoded_signature) do
+      :error ->
+        send_resp(conn, 400, "Bad Request") |> halt()
+
+      {:ok, decoded_signature} ->
+        authenticate_signature(decoded_signature, signed_data, user_public_key, conn)
+    end
+  end
+
+  defp authenticate_signature(signature, message, public_key, conn) do
+    [key_entry] = :public_key.pem_decode(public_key)
+    public_key = :public_key.pem_entry_decode(key_entry)
+
+    is_valid =
+      :public_key.verify(
+        message,
+        :sha256,
+        signature,
+        public_key
+      )
+
+    if is_valid do
+      conn
+    else
+      send_resp(conn, 401, "Unauthorized") |> halt()
+    end
+  end
+
+  defp get_user_name(conn) do
+    user_name = conn.params["user"]
```

Comment:
Could we put required fields in the header? Would this simplify the git hook?
```
Authorization: Signature user="user",signature="base64-payload"
```

Reply:
So the username is already in the JSON body, so it looks like I can effective do the same thing as JWT (and doing what barber said) by refactoring the auth to get the username from the POST body insetad of the query param.

So now...
 - user sends POST,
 - Auth looks at POST, gets the user
 - gets that user's key
 - verify the body against the signature
Which effectively is the same as JWT except we've got an unencoded body instead of an encoded (b64) header.

Reply:
Sounds fine. If we want to explore standards for either the auth header (Signature has an rfc) or JWTs that can be a future iteration on this so we don't scope creep.